Improve Texp_record constructor representation

chambart opened the pull request:
Accessing record information in the typedtree is not very practical, and the information about the type of kept fields (when using the `{ record with label = exp }` construct) is available.

I propose to replace it by:

``` OCaml
  | Texp_record of
      record_label_definition array *
      Types.label_description array *
      Types.record_representation *
       expression option

and record_label_definition =
  | Kept of Types.type_expr
  | Overridden of Longident.t loc * expression
```

This allows to easily access a particular field information (instead of searching for a field with the right field position).

The main motivation for this change is to provide the type information to translcore to propagate it to the lambda. In particular, the PR https://github.com/ocaml/ocaml/pull/336 would make use of it. I had a previous version that needed to do some unification after the type checking, but this version is a lot cleaner.

Notice that even if we lost the order of the fields in the parsetree, this does not affect the evaluation order. I added a test to verify that.

Notice that the evaluation order for record update depend on the number of fields updated:
- if there are less than (size - 3) / 2 updated fields then the evaluation order is from the last index to the first
- otherwise it is the opposite

Would anyone object if I changed the order to avoid this strangeness ?

chambart committed:
Change Texp_record contents to ease the use

It also allows to recover the type of kept fields in a
{ record with field = expr } expression

chambart committed:
Factorisation in the Pexp_record case of Typecore

chambart committed:
Emit type error as soon as we know that a label is missing

chambart committed:
Add test of for record evaluation order

chambart committed:
Test cases for record update evaluation order when using the objdup method

chambart commented:
Merged recent trunk and fix propagation of record update type informations

chambart committed:
Use the type information to annotate updated record allocation

chambart committed:
Test the propagation of updated record type information

chambart committed:
Clean translcore record case using the change in Texp_record

chambart commented:
Re-rebased to the current trunk.

Also use the type information to populate makeblock kinds in case of record update and add some test to check that this effectively allow unboxing in case of record update.

This test was failing first due to a bug introduced by https://github.com/ocaml/ocaml/pull/538 . This contain the fix for it.

chambart commented:
Note also https://github.com/ocaml/ocaml/pull/538 changed the evaluation order of records.
The failure comes from the new evaluation order failing due to this change. If this is considered as irrelevant, I will make every record use the same evaluation order.

chambart commented:
@garrigue @lpw25 or @alainfrisch would you object to this change ?

alainfrisch commented:
I think it's a good change, but I did not review the patch itself.

> record_label_definition array \*   Types.label_description array *

Why not `(Types.label_description * record_label_definition) array`?  This makes an invariant explicit (the two arrays have the same length) and simplify the code that work on them.

Moreover, the comment in Typedtree is confusing since it leads to believe that only overridden fields are listed.  Perhaps just adding a sentence about the fact that all fields of the record type are listed in the arrays would help.

bluddy commented:
I would also recommend switching to inline records where there are more than a pair of parameters, for clarity.

chambart commented:
@alainfrisch, no real reason, it made some change a bit smaller, but this is marginal.

Thanks for the suggestions, I will try this.

chambart committed:
Merge record label descriptions and definition into a single array

chambart commented:
Rererebased...

And applied @alainfrisch suggestions

chambart commented:
@bluddy in that case the name of the type of each field is probably sufficient to describe what is its concern, but that should probably go in another pr soon

alainfrisch commented:
The order I suggested (`Types.label_description * record_label_definition`) seems more natural (it follows the OCaml syntax).

Regarding (inline or not) record for the arguments: if you think this is a good idea, why not doing it in this PR?  Typedtree is processed by external tools (reading .cmt files) so we should indeed aim at user-friendliness and stability (i.e. not change again the definition post 4.04).

chambart committed:
Switch declaration and definition order in Tepx_record fields

chambart committed:
Change Texp_record into an inline record

chambart committed:
Record evaluation order is always the same

The expressions defining a record are evaluated from the last field in
the type definition to the first one.

alainfrisch commented on utils/misc.ml:
```diff
@@ -181,6 +181,26 @@ module Stdlib = struct
       | Some a -> f a
   end
+
+  module Array = struct
```
This helper module is now useless, I think.

chambart committed:
Update Changes

chambart committed:
Remove unused functions from Misc

alainfrisch commented on typing/typedtree.mli:
```diff
@@ -186,9 +186,22 @@ and expression_desc =
+            If the type is { l1: t1; l2: t2 }, the expression
+            { E0 with t2=P2 } is represented as
+            Texp_record
+              ([| l1, Kept t1; l2 Override P2 |], representation,
```
This needs to be updated.

alainfrisch commented:
Good to merge in my opinion (after fixing the doc).

chambart committed:
Update Texp_record documentation

chambart commented:
@alainfrisch thanks for the review.
